elsa/_cli.py

```diff
@@ -11,132 +11,203 @@
-    app.config['SERVER_NAME'] = urllib.parse.urlparse(base_url).netloc
-
-    # make sure Frozen Flask warnings are treated as errors
-    warnings.filterwarnings('error', category=flask_frozen.FrozenFlaskWarning)
-
-    try:
-        freezer.freeze()
-    except flask_frozen.FrozenFlaskWarning as w:
-        print('Error:', w, file=sys.stderr)
-        sys.exit(1)
-
-
-def inject_cname(app):
-    """Create CNAME route for GitHub pages"""
-    @app.route('/CNAME')
-    def cname():
-        return Response(app.config['SERVER_NAME'],
-                        mimetype='application/octet-stream')
+class Elsa:
+    """Elsa freezes websites written in Flask"""
+
+    DEFAULT_PORT = 8003
+    DEFAULT_REMOTE = 'origin'
+
+    def __init__(self, app, *, freezer=None, base_url=None):
+        """Initialize Elsa.
+
+        Arguments:
+
+        * app: and instance of Flask WSGI application
+        * freezer:  flask_frozen.Freezer-like instance (optional)
+        * base_url: URL for the application, used for external links (optional)
+        """
+        self.app = app
+        self.freezer = freezer or ShutdownableFreezer(app)
+        self.base_url = base_url
+
+    def _base_url_help(self, *, freeze=False):
+        freeze = ' with --freeze' if freeze else ''
+        url = self.base_url
+        default = 'default {}'.format(url) if url else 'mandatory' + freeze
+        return 'URL for the application, used for external links, ' + default
+
+    def _inject_cname(self):
+        """Create CNAME route for GitHub pages"""
+        @self.app.route('/CNAME')
+        def cname():
+            return Response(self.app.config['SERVER_NAME'],
+                            mimetype='application/octet-stream')
+
+    def _inject_shutdown(self):
+        """Create a shutdown route"""
+        inject_shutdown(self.app)
+
+    def freeze(self, path=None, base_url=None):
+        """Freeze the app
+
+        Arguments:
+        * path: location of frozen website (tries FREEZER_DESTINATION app
+                config if not provided)
+        * base_url: URL for the application, used for external links
+                    (tries self.base_url and FREEZER_BASE_URL app config
+                    if not provided)"""
+        base_url = (base_url or self.base_url or
+                    self.app.config.get('FREEZER_BASE_URL'))
+        if not base_url:
+            raise ValueError('No base URL provided!')
+        self.app.config['FREEZER_BASE_URL'] = base_url
+
+        path = path or self.app.config.get('FREEZER_DESTINATION')
+        self.app.config['FREEZER_DESTINATION'] = path
+        self.app.config['SERVER_NAME'] = urllib.parse.urlparse(base_url).netloc
+
+        if not path:
+            raise ValueError('No path provided!')
+
+        # make sure Frozen Flask warnings are treated as errors
+        warnings.filterwarnings('error',
+                                category=flask_frozen.FrozenFlaskWarning)
+
+        print('Generating HTML...')
```

Comment:
Should an API like this print on its own?  Perhaps logging as INFO would be better, since a logger is already in use.  Alternatively, a `verbose` parameter.